File: rust/btree.rs
use std::time::Instant;

enum Tree {
    Empty,
    Node { left: Box<Tree>, right: Box<Tree> },
}

fn init_tree(depth: u32) -> Tree {
    if depth == 0 {
        Tree::Empty
    } else {
        Tree::Node {
            left: Box::new(init_tree(depth - 1)),
            right: Box::new(init_tree(depth - 1)),
        }
    }
}

fn node_count(tree: &Tree) -> u32 {
    match tree {
        Tree::Empty => 0,
        Tree::Node { left, right } => 1 + node_count(left) + node_count(right),
    }
}

fn stats(data: &[f32]) -> (f32, f32) {
    let n = data.len() as f32;
    let avg: f32 = data.iter().sum::<f32>() / n;
    let stddev = (data.iter().map(|x| (*x - avg).powf(2.0)).sum::<f32>() / n).sqrt();
    (avg, stddev)
}

fn main() {
    const K: usize = 10;
    let mut times = [[0.0; K]; 22];
    
    for i in 0..K {
        let start = Instant::now();
        for depth in 0..=21 {
            let runs = Instant::now();
            let tree = init_tree(depth);
            let count = node_count(&tree);
            let ends = Instant::now();

            //    println!("Depth {}: Nodes = {}, Runtime = {:?}", depth, count, (ends - runs));
            // println!("{}\t{}\t{}", depth, count, (ends - runs).as_micros());
            
            times[depth as usize][i] = (ends - runs).as_micros() as f32;
            println!("{}", (ends - runs).as_micros());
        }

        let end = Instant::now();
        println!("Took {:?}", (end - start));
    }
    
    println!("Results");
    for i in 0..=21 {
        let (avg, stddev) = stats(&times[i]);
        println!("{}\t{}", avg, stddev);
    }
}

File: ocaml/btree.ml
open Core
type bnode = 
  BNone |
  BNode of {right: bnode; left: bnode}


(** The variants gave me some trouble at first. I had to understand how the 
 variants worked but after that it was easy to define a tree. They are like enums but even
 better **)

let rec init_tree depth =
  if depth = 0 then BNone else BNode {right=init_tree (depth-1); left=init_tree (depth-1)}

let rec node_count = function
  | BNode {right=r; left=l} -> 1+node_count r + node_count l
  | BNone -> 0


let rec benchmark k i = 
  if k <=0 then [] 
  else

    let t1 = Time_ns.now() in
    
    ignore(node_count (init_tree i));

    let t2 = Time_ns.now() in
    (* Stdlib.print_int (Time_ns.to_int_ns_since_epoch t2 - Time_ns.to_int_ns_since_epoch t1); *)
    (Time_ns.to_int_ns_since_epoch t2 - Time_ns.to_int_ns_since_epoch t1)::(benchmark (k-1) i)


let stats data =
    let n = Int.to_float (List.length data) in 
    let avg = ((Int.to_float (BatList.sum data)) /. n) in
    let stddev = ((List.fold_right data ~f:(fun num acc -> acc +. BatFloat.pow ((Int.to_float num)-.avg) 2.) ~init:0.) /. n) |> Float.sqrt in
    (avg, stddev)



let rec main i = 
  if i > 21 then () else
  let (avg, stddev) = stats (benchmark 10 i) in
  Stdlib.print_float avg;
  Stdlib.print_string "\t";
  Stdlib.print_float stddev;
  Stdlib.print_newline ();
  main (i +1)

let () = main 0